chain/chunks/src/lib.rs

```diff
@@ -111,37 +113,34 @@ impl RequestPool {
             switch_to_others_duration,
             switch_to_full_fetch_duration,
             max_duration,
-            requests: HashMap::default(),
+            requests: SizedCache::with_size(MAX_NUM_CHUNK_REQUEST),
         }
     }
-    pub fn contains_key(&self, chunk_hash: &ChunkHash) -> bool {
-        self.requests.contains_key(chunk_hash)
+    pub fn contains_key(&mut self, chunk_hash: &ChunkHash) -> bool {
+        self.requests.cache_get(chunk_hash).is_some()
     }
 
     pub fn insert(&mut self, chunk_hash: ChunkHash, chunk_request: ChunkRequestInfo) {
-        self.requests.insert(chunk_hash, chunk_request);
+        self.requests.cache_set(chunk_hash, chunk_request);
     }
 
     pub fn remove(&mut self, chunk_hash: &ChunkHash) {
-        let _ = self.requests.remove(chunk_hash);
+        self.requests.cache_remove(chunk_hash);
     }
 
     pub fn fetch(&mut self) -> Vec<(ChunkHash, ChunkRequestInfo)> {
-        let mut removed_requests = HashSet::<ChunkHash>::default();
-        let mut requests = Vec::new();
-        for (chunk_hash, mut chunk_request) in self.requests.iter_mut() {
+        let mut requests = vec![];
+        for chunk_hash in self.requests.key_order().cloned().collect::<Vec<_>>() {
```

Comment:
Why do you need to close and collect into Vec here? Can't just iter over the key_order collection?

Reply:
Because key_order holds a reference to the cache and inside the loop I need to use mutable references.